### ISSUE TITLE
Support de-pagination

### ISSUE THREAD

**Issue by mathstuf** [mgattozzi/github-rs#18]
Paginated queries are a bit annoying and the `next` headers given by github itself are better than just doing blind page bumps (race conditions with remote actions), so a way to just handle the pagination internally would be great.

Separate calls for each is probably unweildly, so maybe something like:

```rust
enum Pagination {
    SinglePage(page_num: u64, page_size: u64),
    FirstN(count: usize),
    All,
}
```

Then API rate limiting could also be detected in one place and handled gracefully.

**Comment by mgattozzi** [comment]
I like this idea. I think I originally implemented pagination as it is now to just get it working but the function calls are pretty ugly as is. It'd be nice to handle it with the next headers. I was spending more time working on the actual calls then all the underlying implementation.

**Comment by mgattozzi** [comment]
I'll be looking into this at some point. Probably in a 0.6 release. I want to pass it as part of the headers through a function call (that is now optional) and then figure out how to handle the next headers somehow.

**Comment by mgattozzi** [comment]
@jansol what do you mean by caching exactly? Each call to next() in the iterator would get the next set of pages that were specified. Or am I thinking of something else?

**Comment by jansol** [comment]
I'm thinking that `next()` on the iterator would return one item, not one page. The implementation should fetch the data in reasonably sized pages (say, 25 items per page like the issues page seems to do).

**Comment by mgattozzi** [comment]
Oh I see! Basically query the amount of pages they want, cache them somewhere, and allow them to be iterated over item by item, and not by page.

**Comment by jansol** [comment]
Yup. With this approach the user doesn't have to concern themselves with the page size at all and it allows doing things in a functional style, which is always nice.

**Comment by mathstuf** [comment]
Well, you still want to control page size because if you're getting them all, cutting your queries by a factor of 4 (25 versus the pagemax of 100) is very useful.

**Comment by Michael-F-Bryan** [comment]
> I think I originally implemented pagination as it is now to just get it working but the function calls are pretty ugly as is.

@mgattozzi how exactly is pagination implemented at the moment? I'm currently trying to list all repos starred or owned by a user and am having issues where you only get the first 30 or so repos. Is there an easy way to iterate over the results from a query?

I was imagining some extra `Paginated` trait which will has a default impl that inspects headers and will fetch the next page, yielding each item on the page one by one along the same lines as what @jansol said:

> I'm thinking that next() on the iterator would return one item, not one page. The implementation should fetch the data in reasonably sized pages (say, 25 items per page like the issues page seems to do) and cache it iternally.

The `Paginated` trait might look somewhat like this:

```rust
trait Paginated: Executor {
  fn paginated<T>(self) -> impl Iterator<Item = Result<T, Error>>
  where T: Deserialize,
        Self: Sized
  {
    // inspect headers, caching the current page and fetching the next one every
    // time someone calls `next()`.
  }
}
```

**Comment by mathstuf** [comment]
Pagination is handled by HTTP headers sent by Github. I don't know how much this matters with APIv4 though.

**Comment by mathstuf** [comment]
Note that "all" pagination is likely to hit API limits with large requests, so there may need to be backoff (there are headers for this) in the async case. I don't know what to do for the sync case.

**Comment by mgattozzi** [comment]
API V4 (as I understand it) just sends it all back to you. You can limit how many you want with an input parameter to the calls.

@Michael-F-Bryan Right now pagination is not handled yet. Unfortunately the only workaround (for now) I can think of is the custom query builder executing for an endpoint and building up another request using the pagination link returned and executing that.

I'm open to a PR or suggestions on how to implement this correctly because not hitting API limits is kind of a pain to work with. Paginated as described above seems like a good choice with a possibility of an ApiLimitError being a possible return value in the case of doing to many while iterating.

**Comment by mathstuf** [comment]
[This blog post](http://xion.io/post/code/rust-unfold-pagination.html) has interesting ideas related to this issue.

**Comment by mgattozzi** [comment]
Really wish generators were stable to do this. Either way that's a neat post. Thanks @mathstuf. This seems fairly reasonable and would also help towards #72 working to actually provide an async interface.

**Comment by mgattozzi** [comment]
So this is the completely "Janky Not Nice" version I threw together cause now I needed it for my own stuff. Just wanted to give interested people a chance to look it over and suggest better things because this is in no way robust.
https://github.com/mgattozzi/github-rs/tree/pagination

**Comment by mgattozzi** [comment]
@gdesmott pinging you as well for this issue since you also seem interested in depagination

**Comment by nbigaouette-eai** [comment]
Until this is implemented in a nice way, how should we go about getting multiple pages?

For example, how would one get issues that appear on the second page? The [github docs](https://developer.github.com/v3/guides/traversing-with-pagination/) mentions the `page=X` and `per_page=Y" url arguments but I couldn't find how to control those when using `client.get().repos().owner("owner").repo("repo").issues().execute::<Vec<Value>>()`...

**Comment by nbigaouette-eai** [comment]
I ended up using a custom enpoint like that:
```rust
let issues_endpoint = format!(
        "repos/{}/{}/issues?state=all&per_page={}&page={}",
        owner, repo, per_page, page
    );
```
